---
title: "Analysis report of a fEVE clustering analysis"
author: "Asloudj Yanis"
date: 'Compiled: `r format(Sys.Date(), "%B %d, %Y")`'
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Analysis report of a fEVE clustering analysis}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
bibliography: references.bib
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  message = FALSE,
  warning = FALSE
)
```

# Analysis summary

After conducting a fEVE clustering analysis, a collection of tables is generated. Some functions are readily implemented in the fEVE package to plot the content of the tables. In this vignette, we present these functions. First, we apply the scEVE instance - which is dedicated to single-cell RNA-seq data - on a glioblastoma dataset, to obtain such tables.

```{r feve}
library(feve)
data <- sceve_LoadData("Darmanis_HumGBM")
params <- get_parameters("scEVE")
results <- feve(data$dataset, params=params)
```

### get_plot_summary()

The function `get_plot_summary()` is used to report the hierarchy of clusters (a), as well as the sizes (b) and the number of characteristic features (c) of each leaf cluster. The number of samples and the most characteristic feature of each cluster are also written directly in their respective bars.

```{r get_plot_summary}
get_plot_summary(results$records)
```

### get_plot_compositions()

The function `get_plot_compositions()` leverages a ground truth. It is used to plot the relationships of the predicted clusters (a), as well as the composition of the leaf clusters according to some ground truth (b), and their sizes (c).

```{r get_plot_compositions}
get_plot_compositions(results$records, data$ground_truth)
```

**Note** that the color palette used to represent the ground truth (Kelly) is limited to 22 classes.

### get_plot_resolutions()

The function `get_plot_resolutions()` leverages a ground truth. It is used to plot the quality of the clusters predicted at each resolution, according to four metrics:

-   the `ARI` and the `NMI`, which quantify the quality of a clustering according to its similarity to a ground truth,

-   the `nPurity` and the `SI`, which quantify the quality of a clustering according to the (dis-)similarity of the transcriptomes in and out of the clusters.

For every metric, a higher value indicates a better clustering quality.

```{r get_plot_resolutions()}
p <- get_plot_resolutions(data, results$records$samples, method_label="scEVE")
print(p)
```

### get_plot_contributions()

The composite plot generated with `get_plot_contributions()` is used to investigate which methods contributed to the prediction of each robust cluster. It is composed of an upper barplot where the robustness of each cluster is reported, and a lower heatmap where the cell `x,y` is colored if the method `x` helped predict the cluster `y`.

```{r get_plot_contributions()}
get_plot_contributions(results$records)
```

### write_analysis_report.R

You can call the function `write_analysis_report()` to save a copy of each plot presented in this vignette, in a single .pdf file.

```{r get_analysis_report()}
write_analysis_report(data$dataset, results$records, data$ground_truth)
```

### Miscellaneous

Multiple functions presented in this vignette require ground truths: e.g. `get_plot_resolutions()`, `get_plot_compositions()`. If you are investigating novel datasets, make sure to define a placeholder ground truth before calling these functions.

If the function `get_analysis_report()` is called without a ground truth, a placeholder ground truth is generated automatically by randomly shuffling the cluster labels predicted by fEVE.

As a result, on the .pdf file, the colored bars generated by `get_plot_compositions()` will likely be very heterogeneous, but very similar to each other. Also, the fEVE clustering performances shown with `get_plot_resolutions()` will likely be extremely low - for `ARI` and `NMI` - or extremely high - for `nPurity` and `SI` - compared to the random ground truth.

# Session information

```{r sessioninfo}
library(sessioninfo)
sessioninfo::session_info()
```
